### PR TITLE
Track model

### PR DESCRIPTION
**Pull request by AlecRosenbaum**
* add TrainModel integration where possible. Still waiting on a few functions.
* add force majeure functionality within occupancy and comm code
* add ability to set force majeure block status within GUI
* add "Set Operational" force majeure option within GUI

### CODE REVIEW

File: src/main/java/trackmodel/TrackModel.java [hunk 1]
```diff
@@ -330,11 +332,25 @@ protected void clearDB() {
      *
      * @param      blockId  The block identifier
      *
-     * @return     True if occupied, False otherwise.
+     * @return     True if occupied (or broken), False otherwise.
      */
     public boolean isOccupied(int blockId) {
         this.update();
 
+        switch (this.getStatus(blockId)) {
```
Comment (iggy14750) [inline src/main/java/trackmodel/TrackModel.java:340]: Did you forget that Profetta told us that WC/TkM do not see the repair status of the track? If anything is wrong with it, it is considered occupied.

Comment (AlecRosenbaum) [reply src/main/java/trackmodel/TrackModel.java:340]: Thats what this code is doing...

**Review by iggy14750 — CHANGES_REQUESTED** [review]
TrackModel does not have information about the repair state of the track. You can't ask the track what's wrong with it. CTC can keep track of it, but not us.

File: src/main/java/trackmodel/TrackModel.java (hunk 1 above)
Comment (iggy14750) [reply src/main/java/trackmodel/TrackModel.java:340]: getStatus is still a public method, yes?

Comment (AlecRosenbaum) [reply src/main/java/trackmodel/TrackModel.java:340]: The track model does have to keep track of when a block in being repaired. The CTC still has to notify the track that it is being repaired, it just can read the track status. 
GetStatus is public, but probably should be protected. 
SetStarus needs to remain public. 
All I changed here is to have the track model actually react to the status of the block

Comment (AlecRosenbaum) [reply src/main/java/trackmodel/TrackModel.java:340]: Can't read the track status*

Comment (iggy14750) [reply src/main/java/trackmodel/TrackModel.java:340]: Yeah, and I'm pretty sure Profetta would tell you that even using the class BlockStatus, even without any external interface to access it (which you do still have), he would say that means you know more than you should know. He talked about having a dependency on BlockStatus in subsystem review.

**Comment by iggy14750** [comment]
I suggest `break(blockId)`, and `fix(blockId)` methods, which would just change occupancy.

**Comment by mwm29** [comment]
My code still uses the getStatus call so it will fail to build if you remove it. I can change it and make a pull request but I'm in class now

**Comment by AlecRosenbaum** [comment]
No. It is important for me to keep track of block status. Your suggestion would not work with how I keep track of occupancy.

**Comment by AlecRosenbaum** [comment]
Expanding on previous comments (now that I'm not on mobile)...

As per our discussion with Profeta in class, no one is allowed to ask the track model what the status of a block is. However, the track model **must** be aware of the state of it's blocks, and must be notified when a block is being repaired. His statement in class was something along the lines of "in real life the track would know it's being repaired because someone would be physically there repairing it" and he explicitly stated that the CTC still must notify the track model when a block is being repaired and when it has been fixed.

These changes do not address how the CTC notifies the track model (which it does directly, not through the wayside). These changes address taking the block status/force majeure failures into account when returning occupancy or communicating authority and speed.

Additionally, occupancy is recalculated (from scratch) every update cycle. Simply toggling occupancy with `break(id)` and `fix(id)` would not work, as occupancy is cleared and rewritten every update cycle.

What would be possible is to change `setStatus(id, BlockStatus)` to protected, then add public methods for `setRepair(id)` and `setOperational(id)` that simple call the protected `setStatus` method with the correct BlockStatus enum, though this doesn't seem  like a good enough reason to reject this PR

File: src/main/java/trackmodel/TrackModel.java (hunk 1 above)
Comment (iggy14750) [reply src/main/java/trackmodel/TrackModel.java:340]: OK, this isn't the PR to block for this, for sure, but I think we need to get rid of `getStatus` at least at some point.

**Review by iggy14750 — APPROVED** [review]

File: src/main/java/trackmodel/TrackModel.java (hunk 1 above)
Comment (AlecRosenbaum) [reply src/main/java/trackmodel/TrackModel.java:340]: I agree with that. I'll create an issue and be sure to add it to my next PR

Comment (AlecRosenbaum) [reply src/main/java/trackmodel/TrackModel.java:340]: #83

**Pull request merged**